### ISSUE TITLE
@types/jasmine toHaveBeenCalledWith infers parameters for overloads incorrectly

### ISSUE THREAD

**Issue by devversion** [DefinitelyTyped/DefinitelyTyped#42455]
[The function signature for `toHaveBeenCalledWith`](https://github.com/DefinitelyTyped/DefinitelyTyped/commit/7456aa4eec4bcfc03259a7d32e619d0607bbed3b#r37354130) unfortunately breaks if a generic function that consists of overloads is passed. TypeScript is not able to infer the proper overload for functions through generics. Only a single overload signature is considered for parameter inferring. For example:

```ts
function hello(idx: number): void;
function hello(item: object): void;
function hello(arg: any): void { /* impl */}

// This won't work because `MatchableArgs` only infers `object` as parameter type.
expect(hello).toHaveBeenCalledWith(1);
```

This is a [known limitation in TypeScript](https://github.com/microsoft/TypeScript/issues/32418), but what I'd love to see here is a way to work around this limitation. Would it be helpful to provide a generic to `toHaveBeenCalledWith` that can be used to specify parameters manually if needed?

Authors: @zvirja @djungowski @kolodny

**Comment by chivesrs** [comment]
Yup I just saw this happen in our project as well, especially for the `ts3.1/` typings.
I think the right thing to do here is to unblock users from using those typings, even if it means loosening them.

Here's a repro:

```
describe("handles overloaded functions in spies", () => {
    class MyClass {
        myFunc(x: number): number;
        myFunc(x: string, y: string): number;
        myFunc(...p: any[]): number {
            return 4;
        }
    }

    it("works for one parameter overload", () => {
        const s = new MyClass();
        spyOn(s, 'myFunc');
        s.myFunc(4);
        expect(s.myFunc).toHaveBeenCalledWith(4);
    });

    it("works for two parameter overload", () => {
        const s = new MyClass();
        spyOn(s, "myFunc");
        s.myFunc("4", "5");
        expect(s.myFunc).toHaveBeenCalledWith("4", "5");
    });
});
```

This passes in the tests for the outer level jasmine typings, but fails on the `ts3.1/` ones.

I'm going to look into making them less restrictive.

**Comment by chivesrs** [comment]
I go as far as to say, this is a bug in the typings if it means users are getting type errors when they shouldn't.

**Comment by Maximaximum** [comment]
I've opened a similar bug report (#43989), but after a more thorough search I found this one.

This issue is really annoying and it prevents us from updating to the latest `@types/jasmine`, we're stuck with v. 3.3.0. It seems to be quite difficult to even apply a reasonable workaround to this issue, even if we wanted to.

I'm not sure how this issue should be addressed properly, but something really needs to be changed.

As a last resort, IMHO it would be OK even to revert back to `any[]`, as @devversion suggested. `any[]` is at least usable, whereas the current typings aren't.

**Comment by chivesrs** [comment]
I had a PR open against Jasmine and the definition owners decided against making the change.

The current workaround is to use `NonTypedSpyObj`

See https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jasmine/ts3.1/jasmine-tests.ts#L1762

**Comment by RobARichardson** [comment]
This is a bummer. Was trying to upgrade to Jasmine 3.5 so we could use the new propertyNames support when using jasmine.createSpyObj. Will likely have to revert back.

**Comment by mlibby** [comment]
At the moment, the code in https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/jasmine/ts3.1/index.d.ts sits as follows and is definitely broken when used against Jasmine 3.5.

    interface FunctionMatchers<Fn extends Func> extends Matchers<any> {
        toHaveBeenCalledWith(...params: MatchableArgs<Fn>): boolean;

        /**
         * Add some context for an expect.
         * @param message - Additional context to show when the matcher fails.
         */
        withContext(message: string): FunctionMatchers<Fn>;

        /**
         * Invert the matcher following this expect.
         */
        not: FunctionMatchers<Fn>;
    }

This is the only `toHaveBeenCalled...` method left inside `FunctionMatchers<T>` at this point, and it really ought to be removed and the declaration from `Matchers<T>` allowed to stand on its own without being overridden. When running my tests I get the result: `error TS2554: Expected 2 arguments, but got 1.`

If I comment out the offending declaration from `FunctionMatchers<T>` in favor of the one from `Matchers<T>`, everything works as expected.

**Comment by chivesrs** [comment]
@mlibby Can you give a minimal repro of this?

**Comment by ttutisani** [comment]
I faced the same issue as initially described in this thread. Here is the workaround that I used to assert the arguments passed to the function (assuming that is what you all are also trying to do):

```typescript
expect(objectWithMethod.methodToCall.calls.argsFor(1)).toEqual([expectedArg01 as any]);
```

**Comment by RLovelett** [comment]
[I think this is a "bug" in TypeScript itself](https://github.com/microsoft/TypeScript/issues/26591).

The "this" here being inference only handles one overload as reported in the initial report by @devversion [here](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/42455#issue-566897310).

**Comment by AnnSasha** [comment]
> I faced the same issue as initially described in this thread. Here is the workaround that I used to assert the arguments passed to the function (assuming that is what you all are also trying to do):
> 
> ```ts
> expect(objectWithMethod.methodToCall.calls.argsFor(1)).toEqual([expectedArg01 as any]);
> ```

It works for me, you really save my day!

Thanks!
